Add command to make a connection pooling attachment

coreypurcell opened the pull request:
Add command to create an attachment to a DB that uses the URL for pgbouncer. This is to avoid needing something like a build pack to create the config vars with the pgbouncer URL. 

Work to be completed:

* [x] more tests to cover the case where a credential is specified
* [x] query the api to ensure the DB actually has PGbouncer installed

coreypurcell committed:
add command to make a connection pooling attachment

msakrejda commented:
Thanks! At a glance, this looks fine, but I'ld like to give it a more thorough review. I'll try to do that by Monday.

msakrejda commented on commands/connection_pooling.js:
```diff
@@ -0,0 +1,73 @@
+      cli.action.done(`done, v${releases[0].version}`)
+    })
+  )
+}
```
So this is largely copied from [addons:attach](https://github.com/heroku/heroku-cli-addons/blob/master/commands/addons/attach.js), right? Do you think this is the right approach @jdxcode? The code looks fine to me, but I'm nervous about the copy/pasting (though I don't really have a better suggestion at the moment).

coreypurcell replied:
it is very much just copied from addons:attach.

jdx replied:
all seems fine to me!

msakrejda reviewed (commented):
Sorry about the delay. Looks reasonable though I'd love to hear from @jdxcode. Approval pending the checkbox items in the pull request description.

coreypurcell committed:
check if pgbouncer is enabled

coreypurcell committed:
fix styling

coreypurcell committed:
remove the changes to util

coreypurcell committed:
fix styling

binarycleric committed:
Refactoring command to hit shogun for connection pooling checks

It is better to hit shogun to make sure everyone in pgbouncer land is good
before trying to attach the addon. More server-side logic and less worrying
about CLI updates.

binarycleric committed:
Just using the connection-pooling creds endpoint

The addon is attached server-side.

binarycleric committed:
Moving credential to body and out of uri

binarycleric committed:
test cleanup, dropping unnecessary tests

binarycleric committed:
formatting cleanup

binarycleric committed:
Adding --as flag so users can set an expected name

msakrejda commented on commands/connection_pooling.js:
```diff
@@ -0,0 +1,55 @@
+
+  let db = yield fetcher.addon(app, args.database)
+  let addon = yield heroku.get(`/addons/${encodeURIComponent(db.name)}`)
+  let credential = `${flags.credential || 'default'}`
```
This works but the `flags.credentials` is going to be a string anyway, so you could do `flags.credential || 'default'` without wrapping in a template string.

msakrejda commented on commands/connection_pooling.js:
```diff
@@ -0,0 +1,55 @@
+
+  let attachment = yield cli.action(
+    `Enabling Connection Pooling${credential === 'default' ? '' : ' for credential ' + cli.color.addon(credential)} on ${cli.color.addon(addon.name)} to ${cli.color.app(app)}`,
+    heroku.post(`/client/v11/databases/${db.name}/connection-pooling`, {
```
Why not `encodeURIComponent(db.name)` here? I mean, I don't think it really matters--I don't believe add-on names support any URL-unsafe characters--but we should be consistent.

msakrejda commented on test/commands/connection_pooling.js:
```diff
@@ -0,0 +1,66 @@
+      return cmd.run({app: 'myapp', args: {database: 'postgres-1'}, flags: {credential: name}})
+      .then(() => expect(cli.stdout, 'to equal', ``))
+      .then(() => expect(cli.stderr, 'to contain', 'Setting HEROKU_COLOR config vars and restarting myapp... done, v0\n'))
+    })
```
Can you add tests for no credential specified so we can validate it defaults to the right thing and for the `--as` flag present, to validate that's passed along correctly?

binarycleric replied:
@uhoh-itsmaciek Done.

msakrejda reviewed (changes requested):
Some minor comments, but looks great. Sorry it took me so long to follow up. Feel free to ping me in the future.

binarycleric commented:
Will do. Thanks for the feedback @uhoh-itsmaciek.

binarycleric committed:
code cleanup. being consistent

binarycleric committed:
Adds tests for connection-pooling flags

msakrejda commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "heroku-pg",
   "description": "Heroku CLI plugin to manage Postgres.",
-  "version": "2.4.2",
+  "version": "2.5.0",
```
The version should be bumped separately and that's done automatically by `np`: see https://github.com/heroku/heroku-pg#publishing

msakrejda reviewed (changes requested):
Awesome, thanks! Just drop the `version` change and we're good to go.

binarycleric committed:
reverting version
